Mage.Sets/src/mage/cards/x/XanatharGuildKingpin.java

```diff
@@ -0,0 +1,196 @@
+                TargetController.YOU, false
+        );
+        ability.addEffect(new XanatharGuildKingpinLookAtTopCardOfTargetPlayerLibraryAnyTimeEffect());
+        ability.addEffect(new PlayTheTopCardOfTargetPlayerLibraryEffect(new FilterCard("play cards"), true));
+        ability.addTarget(new TargetOpponent());
+        this.addAbility(ability);
+
+
+    }
+
+    private XanatharGuildKingpin(final XanatharGuildKingpin card) {
+        super(card);
+    }
+
+    @Override
+    public XanatharGuildKingpin copy() {
+        return new XanatharGuildKingpin(this);
+    }
+}
+
+class XanatharGuildKingpinRuleModifyingEffect extends ContinuousRuleModifyingEffectImpl {
+
+    public XanatharGuildKingpinRuleModifyingEffect() {
+        super(Duration.EndOfTurn, Outcome.Benefit);
+        staticText = "choose target opponent. This turn, that player can't cast spells";
+    }
+
+    private XanatharGuildKingpinRuleModifyingEffect(final mage.cards.x.XanatharGuildKingpinRuleModifyingEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public mage.cards.x.XanatharGuildKingpinRuleModifyingEffect copy() {
+        return new mage.cards.x.XanatharGuildKingpinRuleModifyingEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        return true;
+    }
+
+    @Override
+    public String getInfoMessage(Ability source, GameEvent event, Game game) {
+        Player targetPlayer = game.getPlayer(getTargetPointer().getFirst(game, source));
+        MageObject mageObject = game.getObject(source.getSourceId());
+        if (targetPlayer != null && mageObject != null) {
+            return "This turn you can't cast spells" +
+                    " (" + mageObject.getLogName() + ')';
+        }
+        return null;
+    }
+
+    @Override
+    public boolean checksEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.CAST_SPELL;
+    }
+
+    @Override
+    public boolean applies(GameEvent event, Ability source, Game game) {
+        return event.getPlayerId().equals(getTargetPointer().getFirst(game, source));
+    }
+}
+
+class XanatharGuildKingpinLookAtTopCardOfTargetPlayerLibraryAnyTimeEffect extends ContinuousEffectImpl {
+
+    public XanatharGuildKingpinLookAtTopCardOfTargetPlayerLibraryAnyTimeEffect() {
+        super(Duration.EndOfTurn, Layer.PlayerEffects, SubLayer.NA, Outcome.Benefit);
+        staticText = "Until end of turn, you may look at the top card of target player's library any time.";
+    }
+
+    private XanatharGuildKingpinLookAtTopCardOfTargetPlayerLibraryAnyTimeEffect(final mage.cards.x.XanatharGuildKingpinLookAtTopCardOfTargetPlayerLibraryAnyTimeEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        if (game.inCheckPlayableState()) { // Ignored - see https://github.com/magefree/mage/issues/6994
```

Comment:
If you copy paste ability then make comment about it, so it can be found later.

Reply:
It's about `PlayTheTopCardOfTargetPlayerLibraryEffect.java`

Reply:
Alright will do!
I wasn't even sure if it was fine to create a new "public" effect or if it would be better to just add it as a "private" one in the CardFile.
Do you think this maybe makes more sense here?

Reply:
In most cases it must be card specific inner ability like CardNameAbilityNameOrEffect. But that PlayTheTopCardOfTargetPlayerLibraryEffect can be useful later, so it can be as public class.

Reply:
But you must use PlayTheTopCardOfLibraryTargetEffect As name or like that. Name notation like xxxTargetEffect at the end used for targeting abilities.